doc/KUKSA.val_data_broker/authorization_alt_d.md

```diff
@@ -0,0 +1,358 @@
+| sub       | REQUIRED - as defined in Section 4.1.2 of [RFC7519]. In cases of access tokens obtained through grants where a resource owner is involved, such as the authorization code grant, the value of "sub" SHOULD correspond to the subject identifier of the resource owner. In cases of access tokens obtained through grants where no resource owner is involved, such as the client credentials grant, the value of "sub" SHOULD correspond to an identifier the authorization server uses to indicate the client application. See Section 5 for more details on this scenario. Also, see Section 6 for a discussion about how different choices in assigning "sub" values can impact privacy. |
+| client_id | REQUIRED - as defined in Section 4.3 of [RFC8693].   |
+| iat       | REQUIRED - as defined in Section 4.1.6 of [RFC7519]. This claim identifies the time at which the JWT access token was issued. |
+| jti       | REQUIRED - as defined in Section 4.1.7 of [RFC7519]. |
+
+#### Authorization claims
+By specifying a scope when requesting an access token from the authorization server,
+the resource owner can limit what a client (e.g. a third party app) is allowed to do.
+
+> If an authorization request includes a scope parameter, the corresponding issued
+> JWT access token SHOULD include a "scope" claim as defined in Section 4.2 of [RFC8693].
+
+> All the individual scope strings in the "scope" claim MUST have meaning for the
+> resources indicated in the "aud" claim. See Section 5 for more considerations about
+>  the relationship between scope strings and resources indicated by the "aud" claim.
+
+| Claim     | Description                                          |
+|-----------|------------------------------------------------------|
+| scope     | The value of the scope claim is a JSON string containing a space-separated list of scopes associated with the token, in the format described in Section 3.3 of [RFC6749] |
+
+### Scope
+
+Access to VSS entries / signals are limited by specifying a scope in the access token provided to
+KUKSA.VAL.
+
+#### Format
+
+The scopes used in KUKSA.VAL are "prefix scopes" that consist of a VSS path, prefixed
+with an action.
+
+| Scope                 | Description                 |
+|-----------------------|-----------------------------|
+|`<ACTION>:<PATH>`      | Allow ACTION for PATH       |
+|`!<ACTION>:<PATH>`     | Deny ACTION for PATH        |
+
+All "deny" scopes have priority over any "allow" scope.
+
+Examples of KUKSA.VAL scope strings:
+
+| Scope string             | Access                                        |
+|--------------------------|-----------------------------------------------|
+|`read:Vehicle.Speed`      | Client can only read `Vehicle.Speed`          |
+|`read:Vehicle.ADAS`       | Client can read everything under Vehicle.ADAS |
+|`actuate:Vehicle.ADAS`    | Client can actuate all actuators under Vehicle.ADAS |
```

Comment:
I think this is interesting, what is a good initial set of scope-Actions-

`read` surely is. We might in the future want to add some more granular read rights as discussed below, but probably some for of general read permission is good

`actuate` Seems ok, but I am a little on the fence here:  So my understanding is we want to express, that we can write target value. the intent is to actuate,  but that does not mean "actuation" really happens, because there is a stack of software below that will be the judge of that. In that case write:target_value would be more exact

`provide`I do not see as a good term. Our terminology doc still pending,  let me just base this on "unwritten truths". So I assume we do use the term "Provider" for what has been a called a feeder so far. I would use the term provider for everything that connects to a "real" car, so also simple VSS services, i.e. a module that listens for some target value for opening a door, and then doing it. (for that he would not need a provide claim, just a read, so maybe if we don't want something along the lines of "write-current", this shall rather be feed.


 another way to go about this, would be, to take a look at the "Using "access_mode" + "fields" for more granularity" section, and define that first, and treat the comon things as shortcut, just examples (not fully thought through). The full form of valid scopes may be

 - `read:current:Vehicle.Body.Trunk`
 - `read:target:Vehicle.Body.Trunk`
 - `read:meta:Vehicle.Body.Trunk`
 - `write:current:Vehicle.Body.Trunk`
 - `write:target:Vehicle.Body.Trunk`
 - `write:meta:Vehicle.Body.Trunk`
 - `write:current:Vehicle.Body.Trunk`
While you would also allow
 - `read::Vehicle.Body.Trunk` (read everything there is about this tree/signal)
 - `actuate:Vehicle.Body.Trunk` could be defined as `actuate` being a shortcut for `write:target`
 - `feed:Vehicle.Body.Trunk` could be defined as `feed` being a shortcut for `write:current`

Reply:
I think it would be good to specify if you get `read` access for free if you have `write`-access (or provide/actuate-access), or if you need to have explicit `read` access as well.

Reply:
I think the general idea with `read`, `actuate` and `provide` is that they provide a "higher abstraction level" than read / write access to individual fields.

There are a couple of benefits to this.

For one, these actions convey _what_ the client is allowed to do, as opposed to the technical details of what that currently means in practice (i.e. write current value, read actuator target etc). And I agree with @erikbosch that this should be better specified.

#### `read`
There seems to be general consensus around what `read` should mean, i.e. someone with this scope should be allowed read everything for the specified paths. If more granularity is needed, a more specific way to describe that can be added.

#### `actuate`
For `actuate`, I think it would be natural if anyone allowed to actuate something would also be allowed to monitor the actuator as it moves along (i.e. read the value). It also seems natural that they are allowed to read the actuator target and metadata, so in short, `read` should be included in `actuate`. If more granularity is needed, a more specific way to describe that can be added.

> So my understanding is we want to express, that we can write target value. the intent is to actuate, but that does not mean "actuation" really happens, because there is a stack of software below that will be the judge of that.

Yes, the intent of `actuate` is that a client with this scope is allowed to actuate. I fail to see how what you are describing is relevant in terms of either intent or access rights. Sorry.

> In that case write:target_value would be more exact

Yes, this would be more "exact", if by that you mean it would more exactly mirror the current implementation of `kuksa.val.v1`. But I can easily imagine a future where KUKSA.VAL can return a response indicating success or failure when a client calls `actuate(..)` or equivalent (see my other way to long design doc :) ). Then `write:target_value` doesn't make sense anymore.

In contrast, I think `actuate` makes sense today, and would continue to make sense as long as it covers the same intent. These improvements in KUKSA.VAL are definitely not guarantied to happen, but by having "scope actions" that correspond to the "intent", this would be a non-issue in terms of authorization at least.

And as described above, specifying it like `write:target_value` wouldn't include the other access rights that should be included part of being allowed to actuate.

#### `provide`
> So I assume we do use the term "Provider" for what has been a called a feeder so far. I would use the term provider for everything that connects to a "real" car, so also simple VSS services, i.e. a module that listens for some target value for opening a door, and then doing it. (for that he would not need a provide claim, just a read, so maybe if we don't want something along the lines of "write-current", this shall rather be feed.

Again, I think this line of thinking is very tightly coupled to the current implementation of KUKSA.VAL. As an example, I think it is highly desirable if providers of the type you describe would have the ability to "present" themselves as providers to KUKSA.VAL instead of "anonymously" subscribing to `actuator_target`, as this would enable a number of features in KUKSA.VAL. One can also imagine a way for these providers to provide feedback of whether the actuation was successful or not (again see the other way to long "design doc" :) ).

The point is, the access rights needed for those things would be included in `provide`, even though we don't even have that functionality today. And those things would probably not map to any specific fields.

I don't see how your description of a provider that happens to not publish any values is
anything which makes `provide` less useful or misleading today.

#### Alternative for fields
> another way to go about this, would be, to take a look at the "Using "access_mode" + "fields" for more granularity" section, and define that first, and treat the comon things as shortcut, just examples (not fully thought through). The full form of valid scopes may be
> * `read:current:Vehicle.Body.Trunk`
> * `read:target:Vehicle.Body.Trunk`
> * `read:meta:Vehicle.Body.Trunk`
> * `write:current:Vehicle.Body.Trunk`
> * `write:target:Vehicle.Body.Trunk`
> * `write:meta:Vehicle.Body.Trunk`
> * `write:current:Vehicle.Body.Trunk`
>
> While you would also allow
> * `read::Vehicle.Body.Trunk` (read everything there is about this tree/signal)
> * `actuate:Vehicle.Body.Trunk` could be defined as `actuate` being a shortcut for `write:target`
> * `feed:Vehicle.Body.Trunk` could be defined as `feed` being a shortcut for `write:current`

Yes, this also crossed my mind, and I think the design of this "escape hatch" of specifying more granular access rights with access mode per field could be useful. But I think it belongs in the same pot as the "deny" scopes. It could be useful, but we should think it through before we include it. The proposed uses of these, I think focus way to much on _how_ a client is allowed to do something opposed to _what_ they are allowed to do. I think that's my main concern, not the ability to limit access to fields.

So the bottom line is that I think these "higher level abstraction" that cover _what_ the client is allowed to do, would serve us best both in terms of:

* **Ease of use**

  I think it's more intuitive with
  
  `"actuate:Vehicle.Body.Trunk"` & `"provide:Vehicle.Body.Trunk"`
  
  than
  
  `"write:current:Vehicle.Body.Trunk"` & `"write:target:Vehicle.Body.Trunk"`.
  
  And I actually had to double check that it wasn't `"write:value:Vehicle.Body.Trunk"` or `"write:actuator_target:Vehicle.Body.Trunk"` or `"write:target_value:Vehicle.Body.Trunk"`.

* **Future proof**

  By not coupling the access rights super tight to the current implementation of KUKSA.VAL, (or the current version of `kuksa.val.v1`), we have a lot more leeway in making KUKSA.VAL better going forward without the need to also change the design of the authorization.

Reply:
I agree with the actuate and provide mapping instead of write:target and write:current. We had the same issue again at the Hackathon. There where a lot of questions around this. In summary most of the people understood after explanation what the difference is, but I think actuate and provide makes it more clear. I think the shortcut idea is a very good compromise, but could also lead to more confusion E.g. if one is using the shortcuts and another the more granular description in their specification. And then they look at each other's stuff.